### PR TITLE
[ENH] New widget: Import Documents

### LINKED ISSUE #101: Question: How to import multiple files or multiple directories?
Hello,
  Nice to see that we got a very fresh version.

  Question: Is it possible to load multiple files and multiple directories for Text Mining?
  I tried opening the whole level up, no luck
  I tried compressing on a file.tsv.gz using tar, no luck

  As it is a group of directories with multiple files (one per document) it would be normal to create a Corpus from such structure. If not, what would work?

  I have something like:

```
Groups
+- Group_1
|  +- 1999
|  |  +- file_1.txt (any valid extension)
|  |  +- . . .
|  +- 2000
|  +- . . .
+- Group_2
    +- 1982
    +- 1985
    +- . . .
```

Thanks and Regards

### PR DESCRIPTION
**Pull request by jerneju**
##### Issue
New widget. Fixes #101 

##### Description of changes
Text can now be imported from:
- Ordinary text file (`.txt`)
- Office Open XML (`.docx`)
- OpenDocument (`.odt`)
- Portable Document Format (`.pdf`)
- Extensible Markup Language (XML) (`.xml`)

New requirements:
- pdfminer3k>=1.3.1
- odfpy>=1.3.5 (https://anaconda.org/conda-forge/odfpy)
- docx2txt>=0.6 (https://anaconda.org/conda-forge/docx2txt)
- lxml

If there are more folders, folder names are also included in Corpus file. They are represented as a class variable and under name category.

##### Includes
- [X] Code changes
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by kernc** [comment]
What about .doc, .odt and .pdf files? Readers surely exist?

**Comment by jerneju** [comment]
@kernc: Hopefully yes. Check this out: http://textract.readthedocs.io/en/latest/

File: orangecontrib/text/widgets/owimporttexts.py [hunk 1]
```diff
@@ -0,0 +1,776 @@
...
+        def fnmatch_any(fname, patterns):
+            return any(fnmatch.fnmatch(fname.lower(), pattern)
+                       for pattern in patterns)
+
+        batch = []
+
+        for path in filescanner:
+            if fnmatch_any(path, patterns):
+                reader = Reader.get_reader(path)
+                textdata = reader.read()
```
Comment (kernc) [inline orangecontrib/text/widgets/owimporttexts.py]: I suppose this may fail with arbitrary error, e.g. on malformed PDF.

**Comment by kernc** [comment]
Wouldn't it be reasonable to expose from-directory corpus loading in the scripting API as well?

File: orangecontrib/text/widgets/owimporttexts.py [hunk 2]
```diff
@@ -0,0 +1,776 @@
...
+        if self.currentPath is not None and \
+                os.path.isdir(self.currentPath) and self.recent_paths and \
+                os.path.samefile(self.currentPath, self.recent_paths[0].abspath):
+            self.recent_cb.setCurrentIndex(0)
+        else:
+            self.currentPath = None
+            self.recent_cb.setCurrentIndex(-1)
+        self.__actions.reload.setEnabled(self.currentPath is not None)
+
+    def customEvent(self, event):
```
Comment (jerneju) [inline orangecontrib/text/widgets/owimporttexts.py]: This widget is a modified copy of **Import Images** which is a part of **Image Analytics** add-on. I need to check this.

File: orangecontrib/text/widgets/owimporttexts.py (hunk 1 above)
Comment (jerneju) [inline orangecontrib/text/widgets/owimporttexts.py]: Sure, `try except` block will be added.

File: orangecontrib/text/widgets/owimporttexts.py (hunk 2 above)
Comment (kernc) [inline orangecontrib/text/widgets/owimporttexts.py]: IIUC this custom event handler, as used, prevents running the `start()` method's code as part of widget constructor. Like if `QTimer.singleShot(1, ...)` were used.

**Comment by ajdapretnar** [comment]
I understand this is WIP, but just a note while you're working: some pdfs get rendered nicely, others poorly and some not at all. Please check the reader and how/whether is works. Also there might be some issues with non-ASCII characters, please check and test this as well. Thanks! :)

Oh also:
1. remove 'group' field. Use it for _class_ only if there's more than one folder.
2. remove 'ext' field. Instead, remove the extension from filename to make it nicer to read: file.txt --> file
3. set 'content' to be a text feature. This is urgent, since without it nothing will work.

File: orangecontrib/text/widgets/owimporttexts.py [hunk 3]
```diff
@@ -0,0 +1,643 @@
...
+    ThreadExecutor, FutureWatcher, methodinvoke
+)
+from Orange.widgets.widget import Output
+
+from orangecontrib.text.corpus import Corpus
+from orangecontrib.text.widgets.utils.readers import TextScan
+from Orange.canvas.preview.previewbrowser import TextLabel
+
+
+def prettyfypath(path):
```
Comment (ajdapretnar) [inline orangecontrib/text/widgets/owimporttexts.py]: Just for the sake of clarity, should this be _prettifypath_? :)

**Comment by jerneju** [comment]
@ajdapretnar : Well, **pdfminer3k** may be a little better module to extract texts from _pdf_ format files. However that module does not work on all required platforms.

File: orangecontrib/text/widgets/owimporttexts.py (hunk 3 above)
Comment (jerneju) [inline orangecontrib/text/widgets/owimporttexts.py]: Yes, probably.

**Comment by kernc** [comment]
What about [pdfrw](https://github.com/pmaupin/pdfrw) (seems recent) or other [pure-python alternatives](https://github.com/pmaupin/pdfrw#pure-python)?

**Comment by jerneju** [comment]
@kernc: Thank you, now I tested by myself. **pdfminer3k** can be installed on Windows.

**Comment by kernc** [comment]
Can we also support epub? https://github.com/aerkalov/ebooklib

**Comment by kernc** [comment]
Not _epub_! Its docs link to some page in French. ebooklib.

File: orangecontrib/text/widgets/utils/readers.py [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+        return self.make_text_data()
+
+
+class PdfReader(Reader):
+    ext = ".pdf"
+    def __init__(self, path):
+        super().__init__(path)
+
+    def read_file(self):
+        fp = open(self.path, 'rb')
```
Comment (kernc) [inline orangecontrib/text/widgets/utils/readers.py]: with open.

File: orangecontrib/text/widgets/utils/readers.py [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+        interpreter = PDFPageInterpreter(rsrcmgr, device)
+        extracted_text = ""
+
+        for page in doc.get_pages():
+            interpreter.process_page(page)
+            layout = device.get_result()
+            for lt_obj in layout:
+                if isinstance(lt_obj, LTTextBox) or isinstance(lt_obj, LTTextLine):
+                    extracted_text += lt_obj.get_text()
+        self.content = extracted_text
```
Comment (kernc) [inline orangecontrib/text/widgets/utils/readers.py]: make
```py
text = []
for page in pages:
   text.append(page.get_text())
text = ' '.join(text)
```

File: orangecontrib/text/widgets/utils/readers.py [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+
+class OdtReader(Reader):
+    ext = ".odt"
+    def __init__(self, path):
+        super().__init__(path)
+
+    def read_file(self):
+        odtfile = load(self.path)
+        texts = odtfile.getElementsByType(text.P)
+        self.content = " ".join(teletype.extractText(texts[i]) for i in range(len(texts)))
```
Comment (kernc) [inline orangecontrib/text/widgets/utils/readers.py]:
```py
' '.join(teletype.extractText(t) for t in texts)
```

File: orangecontrib/text/widgets/utils/readers.py (hunk 4 above)
Comment (jerneju) [inline orangecontrib/text/widgets/utils/readers.py]: Fixed.

File: orangecontrib/text/widgets/utils/readers.py (hunk 6 above)
Comment (jerneju) [inline orangecontrib/text/widgets/utils/readers.py]: Yes, sure.

**Comment by kernc** [comment]
Add .html/.xml support via beautifulsoup (`root_node.getText()`) which is already a dependency.

**Comment by ajdapretnar** [comment]
Ok, how about we stop supporting all possible formats and focus on stabilizing and testing this widget? We need it soon and test it thoroughly before the workshop, so as we said: support basic formats and have a solid widget and later add support for n formats. Thanks!

**Comment by ajdapretnar** [comment]
Works fine for xml, too. For now I see no problems with the widget, but it needs thorough testing. Could you add good tests for the widget, please?

Also, I am seeing this for some pdf files.
screen shot 2017-07-24 at 16 07 50

Is this a bug in the reader or something we can fix?

**Comment by jerneju** [comment]
That is a bug in the reader. I can test this document using another pdf library.

**Comment by jerneju** [comment]
Fixed.

**Comment by ajdapretnar** [comment]
There are a couple of issues still pertaining.
1. Widget crashes with I'm loading some data and pressing Cancel in the middle of it. It breaks the entire Python, as a matter of fact.
2. Set _content_ to be a text_feature, not name. You need to set it with set_text_features().
3. When I load Zdravljica files, they seem to render ok in Corpus Viewer, while I am only getting the first line in Data Table. Any idea why this happens?
4. The output of the widget should be Corpus, not Data. Just a matter of renaming it.

**Comment by ajdapretnar** [comment]
5. A crash occurs when connecting it to Scatter Plot. Probably an issue of SP itself, but just mentioning here, if perhaps you can check what's the problem.

Traceback (most recent call last):
  File "/Users/ajda/orange/orange3/Orange/canvas/scheme/widgetsscheme.py", line 832, in process_signals_for_widget
    widget.handleNewSignals()
  File "/Users/ajda/orange/orange3/Orange/widgets/visualize/owscatterplot.py", line 404, in handleNewSignals
    self.sparse_to_dense(self.subset_data))
  File "/Users/ajda/orange/orange3/Orange/widgets/visualize/owscatterplotgraph.py", line 609, in new_data
    self.set_data(data, **args)
  File "/Users/ajda/orange/orange3/Orange/widgets/utils/scaling.py", line 100, in set_data
    self._compute_scaled_data()
  File "/Users/ajda/orange/orange3/Orange/widgets/utils/scaling.py", line 52, in _compute_scaled_data
    self.valid_data_array = ~np.isnan(no_jit)
TypeError: ufunc 'isnan' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''

**Comment by nikicc** [comment]
The above error sounds like a problem with sparse matrices. Are we outputting sparse matrices from Text Import?

**Comment by jerneju** [comment]
@ajdapretnar @nikicc : Scatter Plot fails on empty sparse X matrix. Yes, corpus created by this widget has empty sparse X. It is automatically created.

**Comment by nikicc** [comment]
@jerneju could we then fix this in Scatter Plot?

**Comment by jerneju** [comment]
@ajdapretnar: Thank you for finding new errors. Let's go:
1) Maybe Import Images crashes as well? I shall check this.
2) Fixed.
3) Because of newlines. I can replace them with spaces. Only first line is seen in Table.
4) Fixed.

**Comment by ajdapretnar** [comment]
I think we should. Some folks might find it natural to connect Scatter Plot to Text Import and getting an error is a bad surprise. 😁

**Comment by jerneju** [comment]
@nikicc : Yes, I am planning to fix Scatter Plot.

**Comment by ajdapretnar** [comment]
> Because of newlines. I can replace them with spaces. Only first line is seen in Table.

Hm. I'm not sure. What do you think @nikicc? I like poem for example to be nicely formatted in Corpus Viewer... If it's a poem, it should look like a poem. But it does look weird in a Data Table. Opinions?

**Comment by nikicc** [comment]
> Because of newlines. I can replace them with spaces. Only first line is seen in Table.

I'm not sure about this. In Corpus Viewer it probably looks much better with the newlines? Table is not meant to be a widget for viewing corpuses anyhow.

**Comment by nikicc** [comment]
> Hm. I'm not sure. What do you think @nikicc? I like poem for example to be nicely formatted in Corpus Viewer... If it's a poem, it should look like a poem. But it does look weird in a Data Table. Opinions?

I Would keep newlines for the sake of Corpus Viewer. Maybe Table's view component could be hacked to remove newlines?

**Comment by ajdapretnar** [comment]
True. Then I guess let's go with newlines and embrace them fully. If there's a newline in a text, we should keep it.

**Comment by ajdapretnar** [comment]
p.s. How does tokenization handle newlines?

**Comment by jerneju** [comment]
Scatter Plot issue fixed.

File: orangecontrib/text/widgets/utils/readers.py [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+        return text_data
+
+    @classmethod
+    def get_categories(cls, text_data):
+        categories = []
+        for textdata in text_data:
+            categories.append(textdata.category)
+        return list(set(categories))
+
+def scan(topdir, include_patterns=("*",), exclude_patterns=(".*",)):
```
Comment (nikicc) [inline orangecontrib/text/widgets/utils/readers.py]: How do I pass `include_patterns ` or `exclude_patterns ` from `TextScan` class; i.e. how can the user scan directories and load only `txt` for example?

I this is not possible, why the fuss about this?

File: orangecontrib/text/widgets/utils/readers.py [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+
+    @classmethod
+    def get_reader(cls, path):
+        ext = pathlib.Path(path).suffix
+        return eval("{}Reader(path)".format(capwords(ext[1:])))
+
+    def read(self, replace_white_space=False):
+        try:
+            textdata = self.read_file()
+        except:
```
Comment (nikicc) [inline orangecontrib/text/widgets/utils/readers.py]: The same happens for [webtext corpus](https://github.com/nltk/nltk_data/blob/gh-pages/packages/corpora/webtext.zip). Out of 6 text files, I only got two if TxtReader is in text mode.

File: orangecontrib/text/widgets/owimporttexts.py [hunk 9]
```diff
@@ -0,0 +1,642 @@
...
+    return item
+
+
+class State(enum.IntEnum):
+    NoState, Processing, Done, Cancelled, Error = range(5)
+
+
+class OWImportTexts(widget.OWWidget):
+    name = "Import Texts"
+    description = "Import text document(s) from a directory(s)"
```
Comment (ajdapretnar) [inline orangecontrib/text/widgets/owimporttexts.py]: I also vote for: Import text documents from folders.
Plural is enough and perhaps folder is more intuitive to directory?

File: orangecontrib/text/widgets/owimporttexts.py [hunk 10]
```diff
@@ -0,0 +1,642 @@
...
+        if self.__state == State.Processing:
+            self.pathlabel.setText(prettifypath(arg.lastpath))
+
+    def commit(self):
+        """
+        Create and commit a Corpus from the collected text meta data.
+        """
+        corpus = None
+        if self._text_data:
+            names = ["name", "path", "content"]
```
Comment (ajdapretnar) [inline orangecontrib/text/widgets/owimporttexts.py]: So TextScan retrives documents from a folder and builds a corpus? Perhaps a rename is in order then? Sth like RetrieveText or GetText? But I agree, scripting should be supported, too.

File: orangecontrib/text/widgets/utils/readers.py [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+                              batch=batch))
+        return text_data
+
+    @classmethod
+    def get_categories(cls, text_data):
+        categories = []
+        for textdata in text_data:
+            categories.append(textdata.category)
+        return list(set(categories))
+
```
Comment (jerneju) [inline orangecontrib/text/widgets/utils/readers.py]: Fixed.

File: orangecontrib/text/widgets/utils/readers.py [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+        if batch and self.report_progress is not None:
+            self.report_progress(
+                    namespace(count=len(text_data),
+                              lastpath=text_data[-1].path,
+                              batch=batch))
+        return text_data
+
+    @classmethod
+    def get_categories(cls, text_data):
+        categories = []
```
Comment (jerneju) [inline orangecontrib/text/widgets/utils/readers.py]: Fixed.

File: orangecontrib/text/widgets/utils/readers.py [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+
+            if len(batch) == 10 and self.report_progress is not None:
+                self.report_progress(
+                    namespace(count=len(text_data),
+                              lastpath=text_data[-1].path,
+                              batch=batch))
+                batch = []
+
+        if batch and self.report_progress is not None:
+            self.report_progress(
```
Comment (jerneju) [inline orangecontrib/text/widgets/utils/readers.py]: Fixed.

File: orangecontrib/text/widgets/utils/readers.py [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+        def fnmatch_any(fname, patterns):
+            return any(fnmatch.fnmatch(fname.lower(), pattern)
+                       for pattern in patterns)
+
+        batch = []
+
+        for path in filescanner:
+            if fnmatch_any(path, patterns):
+                reader = Reader.get_reader(path)
+                textdata = reader.read()
```
Comment (jerneju) [inline orangecontrib/text/widgets/utils/readers.py]: Sure.

File: orangecontrib/text/widgets/utils/readers.py [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+
+
+class TxtReader(Reader):
+    ext = ".txt"
+
+    def read_file(self):
+        with open(self.path, 'rb') as f:
+            content = f.read()
+        self.content = content
+        return self.make_text_data()
```
Comment (jerneju) [inline orangecontrib/text/widgets/utils/readers.py]: Yes, fixed.

File: orangecontrib/text/widgets/utils/readers.py [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+            self.content = re.sub('\s+', ' ', self.content)
+        return TextData(name, self.path, self.ext, category, self.content)
+
+
+class TxtReader(Reader):
+    ext = ".txt"
+
+    def read_file(self):
+        with open(self.path, 'rb') as f:
+            content = f.read()
```
Comment (jerneju) [inline orangecontrib/text/widgets/utils/readers.py]: Fixed.

File: orangecontrib/text/widgets/utils/readers.py [hunk 17]
```diff
@@ -0,0 +1,215 @@
...
+    def get_reader(cls, path):
+        ext = pathlib.Path(path).suffix
+        return eval("{}Reader(path)".format(capwords(ext[1:])))
+
+    def read(self, replace_white_space=False):
+        try:
+            textdata = self.read_file()
+        except:
+            textdata = None
+        self.replace_white_space=replace_white_space
```
Comment (jerneju) [inline orangecontrib/text/widgets/utils/readers.py]: Fixed.

File: orangecontrib/text/widgets/owimporttexts.py [hunk 18]
```diff
@@ -0,0 +1,642 @@
...
+
+class State(enum.IntEnum):
+    NoState, Processing, Done, Cancelled, Error = range(5)
+
+
+class OWImportTexts(widget.OWWidget):
+    name = "Import Texts"
+    description = "Import text document(s) from a directory(s)"
+    icon = ""
+    priority = 110
```
Comment (jerneju) [inline orangecontrib/text/widgets/owimporttexts.py]: Fixed.

File: orangecontrib/text/widgets/owimporttexts.py (hunk 9 above)
Comment (jerneju) [inline orangecontrib/text/widgets/owimporttexts.py]: Fixed.

File: orangecontrib/text/widgets/utils/readers.py [hunk 19]
```diff
@@ -0,0 +1,215 @@
...
+TextData.isvalid = property(lambda self: True)
+
+TextDataError = namedtuple(
+    "TextDataError",
+    ["path", "error", "error_str"]
+)
+TextDataError.isvalid = property(lambda self: False)
+
+
+class Reader():
```
Comment (jerneju) [inline orangecontrib/text/widgets/utils/readers.py]: Fixed.

File: orangecontrib/text/widgets/owimporttexts.py (hunk 10 above)
Comment (jerneju) [inline orangecontrib/text/widgets/owimporttexts.py]: Scripting is now supported. Now we need to choose a proper name.

- `TextScan` (current)
- `RetrieveText`
- `GetText`

File: orangecontrib/text/widgets/owimporttexts.py [hunk 20]
```diff
@@ -0,0 +1,642 @@
...
+
+from Orange.widgets import widget, gui, settings
+from Orange.widgets.utils.filedialogs import RecentPath
+from Orange.widgets.utils.concurrent import (
+    ThreadExecutor, FutureWatcher, methodinvoke
+)
+from Orange.widgets.widget import Output
+
+from orangecontrib.text.corpus import Corpus
+from orangecontrib.text.widgets.utils.readers import TextScan
```
Comment (jerneju) [inline orangecontrib/text/widgets/owimporttexts.py]: All right, I am moving there. But are you sure that `text_import.py` is a proper file name? Maybe a name without an underscore?

File: orangecontrib/text/widgets/owimporttexts.py (hunk 10 above)
Comment (jerneju) [inline orangecontrib/text/widgets/owimporttexts.py]: `ImportDocuments` Same name for widget and scripting part.

File: orangecontrib/text/widgets/owimporttexts.py (hunk 20 above)
Comment (jerneju) [inline orangecontrib/text/widgets/owimporttexts.py]: `import_documents.py`

**Comment by nikicc** [comment]
@jerneju the icon for this is in #289.

File: orangecontrib/text/widgets/utils/readers.py (hunk 7 above)
Comment (jerneju) [inline orangecontrib/text/widgets/utils/readers.py]: Fixed.

File: orangecontrib/text/widgets/utils/readers.py [hunk 21]
```diff
@@ -0,0 +1,215 @@
...
+        if self.replace_white_space:
+            self.content = re.sub('\s+', ' ', self.content)
+        return TextData(name, self.path, self.ext, category, self.content)
+
+
+class TxtReader(Reader):
+    ext = ".txt"
+
+    def read_file(self):
+        with open(self.path, 'rb') as f:
```
Comment (jerneju) [inline orangecontrib/text/widgets/utils/readers.py]: `detect_encoding` from _Orange.data.io_

**Comment by ajdapretnar** [comment]
I get an error when loading my articles:
Traceback (most recent call last):
  File "/Users/ajda/orange/orange3-text/orangecontrib/text/widgets/owimportdocuments.py", line 537, in __onRunFinished
    getattr(self.Warning, error)()
  File "/Users/ajda/orange/orange3/Orange/widgets/utils/messages.py", line 114, in __call__
    self.formatted = self.format(*args, **kwargs)
KeyError: "'U'"

**Comment by ajdapretnar** [comment]
Cancel still doesn't work for me.
Traceback (most recent call last):
  File "/Users/ajda/orange/orange3-text/orangecontrib/text/widgets/owimportdocuments.py", line 547, in cancel
    assert self.__pendingTask is not None
AssertionError

**Comment by jerneju** [comment]
@ajdapretnar : First error is now probably fixed.

**Comment by ajdapretnar** [comment]
In readin xml files, I am getting:
Traceback (most recent call last):
  File "/Users/ajda/orange/orange3-text/orangecontrib/text/import_documents.py", line 62, in read
    self.read_file()
  File "/Users/ajda/orange/orange3-text/orangecontrib/text/import_documents.py", line 147, in read_file
    soup = BeautifulSoup(markup.read(), "lxml")
  File "/Users/ajda/anaconda/envs/o3/lib/python3.5/codecs.py", line 321, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x92 in position 16794: invalid start byte

This is however caught by a Warning I assume (607 problems during execution), which is good. However, I'm wondering whether we can fix this at source, since I guess it's just a decoding issue and can be resolved. :)

**Comment by ajdapretnar** [comment]
I was waiting for some progress on the data, when I decided to quit Orange. I got this in the console:
Traceback (most recent call last):
  File "/Users/ajda/anaconda/envs/o3/lib/python3.5/logging/__init__.py", line 1881, in shutdown
    h.flush()
  File "/Users/ajda/anaconda/envs/o3/lib/python3.5/logging/__init__.py", line 964, in flush
    self.stream.flush()
  File "/Users/ajda/orange/orange3/Orange/canvas/application/outputview.py", line 207, in flush
    self.flushed.emit()
AttributeError: 'TextStream' does not have a signal with the signature flushed()

**Comment by ajdapretnar** [comment]
Also, in the PR, I'm constantly seeing some messages in the console, whenever I'm placing or moving widgets around.
INFO:Orange.canvas.scheme.scheme:Added node 'Preprocess Text' to scheme ''.
INFO:Orange.canvas.canvas.scene:Added item '<Orange.canvas.canvas.items.nodeitem.NodeItem object at 0x188f39ca8>' to 'CanvasScene(objectName='', ...)'
INFO:Orange.canvas.scheme.scheme:Added link 'Import Documents' ('Corpus') -> 'Preprocess Text' ('Corpus') to scheme ''.
INFO:Orange.canvas.scheme.signalmanager:Link added (SchemeLink((Import Documents, Corpus) -> (Preprocess Text, Corpus))). Scheduling signal data update.
INFO:Orange.canvas.canvas.scene:Added link 'Import Documents' -> 'Preprocess Text' to 'CanvasScene(objectName='', ...)'
INFO:Orange.canvas.canvas.scene:Setting interaction 'None' to 'CanvasScene(objectName='', ...)'

I don't think this should be displayed in there. Is this some forgotten print somewhere?

**Comment by jerneju** [comment]
@ajdapretnar : Please test all of them except last two. Those two issues occurs because of **pdfminer3k**.

File: orangecontrib/text/import_documents.py [hunk 22]
```diff
@@ -0,0 +1,262 @@
...
+)
+TextData.isvalid = property(lambda self: True)
+
+TextDataError = namedtuple(
+    "TextDataError",
+    ["path", "error", "error_str"]
+)
+TextDataError.isvalid = property(lambda self: False)
+
+log = logging.getLogger()
```
Comment (kernc) [inline orangecontrib/text/import_documents.py]: Add `__name__`

File: orangecontrib/text/import_documents.py [hunk 23]
```diff
@@ -0,0 +1,262 @@
...
+
+
+class Reader:
+    _readers = ["TxtReader", "DocxReader", "OdtReader", "PdfReader", "XmlReader"]
```
Comment (kernc) [inline orangecontrib/text/import_documents.py]: You can make this:
```py
class Reader(metaclass=Orange.util.Registry):
```
The all reader implementations can be iterated over with:
```py
for reader in cls.registry:
    if ext in reader.ext:
        ...
```

File: orangecontrib/text/import_documents.py [hunk 24]
```diff
@@ -0,0 +1,262 @@
...
+            interpreter.process_page(page)
+            layout = device.get_result()
+            for lt_obj in layout:
+                if isinstance(lt_obj, LTTextBox) or isinstance(lt_obj, LTTextLine):
+                    extracted_text.append(lt_obj.get_text())
+        self.content = ' '.join(extracted_text)
+
+
+class XmlReader(Reader):
+    ext = [".xml"]
```
Comment (kernc) [inline orangecontrib/text/import_documents.py]: Just curious, why not html and newspaper?

File: orangecontrib/text/import_documents.py [hunk 25]
```diff
@@ -0,0 +1,262 @@
...
+
+    paths = []
+
+    for dirpath, dirnames, filenames in os.walk(topdir):
+        for dirname in list(dirnames):
+            # do not recurse into hidden dirs
+            if fnmatch.fnmatch(dirname, ".*"):
+                dirnames.remove(dirname)
+
+        def matches_any(fname, patterns):
```
Comment (kernc) [inline orangecontrib/text/import_documents.py]: Move definition outside of loop.

File: orangecontrib/text/widgets/utils/readers.py [hunk 26]
```diff
@@ -0,0 +1,215 @@
...
+    ext = ".xml"
+
+    def read_file(self):
+        with open(self.path) as markup:
+            soup = BeautifulSoup(markup.read(), "lxml")
+        self.content = soup.get_text()
+        return self.make_text_data()
+
+
+DefaultFormats = ("docx", "odt", "txt", "pdf", "xml")
```
Comment (jerneju) [inline orangecontrib/text/widgets/utils/readers.py]: Moved.

File: orangecontrib/text/import_documents.py (hunk 22 above)
Comment (jerneju) [inline orangecontrib/text/import_documents.py]: Added.

File: orangecontrib/text/import_documents.py (hunk 23 above)
Comment (jerneju) [inline orangecontrib/text/import_documents.py]: I like that.

File: orangecontrib/text/import_documents.py (hunk 24 above)
Comment (jerneju) [inline orangecontrib/text/import_documents.py]: I had some issues and `html` is not so high priority at the moment. But no worries, I am planning to implement other formats as well.

File: orangecontrib/text/import_documents.py (hunk 25 above)
Comment (jerneju) [inline orangecontrib/text/import_documents.py]: Fixed.

**Comment by nikicc** [comment]
I only found one small glitch. If I click cancel, the Info box says `cancelled` but old corpus is still on the output. It probably shouldn't be anymore.

**Comment by nikicc** [comment]
Shit, pdfminer3k is not available on conda, at least for linux-64 channel ([see this PR](https://github.com/conda-forge/orange3-text-feedstock/pull/4)). @jerneju is there something else we could use, or should we start working towards putting it on conda-forge?

@ajdapretnar we didn't plan to send installation instructions for conda right?

**Comment by jerneju** [comment]
@nikicc: **PyPDF2** (https://anaconda.org/conda-forge/pypdf2) Firstly I implemented that module but then I moved to **pdfminer3k**. I think the latter is better.

**Pull request merged**